blob/store.go

```diff
@@ -108,11 +109,20 @@ type KVCore interface {
 	// in the store, Delete must report an ErrKeyNotFound error.
 	Delete(ctx context.Context, key string) error
 
-	// List calls f with each key in the store in lexicographic order, beginning
-	// with the first key greater than or equal to start.  If f reports an error
-	// listing stops and List returns.  If f reported an ErrStopListing error,
-	// List returns nil; otherwise List returns the error reported by f.
-	List(ctx context.Context, start string, f func(string) error) error
+	// List returns an iterator over each key in the store greater than or equal
+	// to start, in lexicographic order.  Each pair reported by the iterator
+	// must have a nil error, except possibly the last.
```

Comment:
When I've implemented this shape in the past, I've made a stronger promise to avoid the io.Reader confusion: each loop of the iterator yields either `(T, nil)` or `(zero, err)`, but never `(T, err)`. This is a common footgun when implementing readers, because "An error means other values are meaningless" is such an ingrained behavior. I'd suggest making the same stronger promise here:
 - The iterator must yield a value XOR an error, never both or neither.
 - If the iterator yields an error, the iteration must end after that error yield returns, even if the yield returned true.

Reply:
Thanks, that was actually what I was trying to imply, but your note makes me realize I did not explicitly note that the error must be the _only_ meaningful memeber of its tuple.

Reply:
I adopted substantially that sense, though there is no guarantee that `"", nil` is not a valid key (though some stores don't accept `""`). So I did not constraint it quite exactly that way.